File: 1208-get-equal-substrings-within-budget/1208-get-equal-substrings-within-budget.java
class Solution {
    public int equalSubstring(String s, String t, int maxCost) {
        int N = s.length();
        int[] costs = new int[N];
        int answer = 0;
        
        int left = 0, sum = 0;
        for(int i = 0; i < s.length(); i++) {
            costs[i] = Math.abs(t.charAt(i) - s.charAt(i));
            sum += costs[i];
            if(sum <= maxCost) {
                //System.out.printf("%c(%d)부터 %c(%d)까지 변경 가능 (sum = %d)\n", s.charAt(left),left, s.charAt(i), i, sum);
                answer = Math.max(answer, i - left + 1);
            } else {
                sum -= costs[left++];
            }
        }
        
        return answer;
    }
}

File: 1208-get-equal-substrings-within-budget/1208-get-equal-substrings-within-budget.py
class Solution:
    def equalSubstring(self, s: str, t: str, maxCost: int) -> int:
        left, costsum, answer = 0, 0, 0
        costs = []
        for i, (x, y) in enumerate(zip(s, t)):
            costs.append(abs(ord(x) - ord(y)))
            costsum += costs[i]
            
            if costsum <= maxCost:
                answer = max(answer, i - left + 1)
            else:
                costsum -= costs[left]
                left += 1
        return answer
        